### PR TITLE
fix(migrations): drop the duplicate 013 pair and guard against recurrence

### PR DESCRIPTION
**Pull request by jaredglaser**
## What and why

`migrations/` carried two files numbered `013` (`013_drop_agent_token_columns.sql` and `013_stack_status.sql`). The filenames differ, so git saw no conflict when the two PRs landed a day apart, and nothing in CI checked for it. This deletes the `013_stack_status.sql` / `019_drop_stack_status.sql` pair, which cancels exactly, and adds a test that fails on any future duplicate prefix.

### The pair cancels, with evidence

`013_stack_status.sql` creates exactly three objects; `019_drop_stack_status.sql` drops exactly those three with `IF EXISTS`:

| created by 013 | dropped by 019 |
| --- | --- |
| `CREATE TABLE stack_status` | `DROP TABLE IF EXISTS stack_status` |
| `CREATE OR REPLACE FUNCTION notify_stack_change()` | `DROP FUNCTION IF EXISTS notify_stack_change()` |
| `CREATE TRIGGER stack_status_notify ON stack_status` | `DROP TRIGGER IF EXISTS stack_status_notify ON stack_status` |

Residue checked and found absent:

- **No sequences.** `stack_status` has no `SERIAL` or `IDENTITY` column; its columns are `stack TEXT`, `host TEXT`, `containers JSONB DEFAULT '[]'`, `updated_at TIMESTAMPTZ DEFAULT now()`.
- **No orphan indexes or constraints.** The only index is the implicit `PRIMARY KEY (stack, host)` index, dropped with the table. Same for the FK constraint.
- **No other migration depends on it.** `grep -rn "stack_status\|notify_stack_change" migrations/` matches only `013_stack_status.sql` and `019_drop_stack_status.sql`. Migrations 014 through 018 never reference the table, so nothing between the create and the drop breaks.
- **`notify_stack_change()` is not shared.** No other migration creates a trigger using it, so 019's unconditional `DROP FUNCTION` removes something only 013 created.
- **The `managed_hosts` foreign key changes nothing about `managed_hosts`.** `stack_status.host` was `REFERENCES managed_hosts(name) ON DELETE CASCADE`. The uniqueness that FK requires comes from `009_deploy_pipeline.sql`, which declares `name TEXT UNIQUE NOT NULL` on `managed_hosts` independently; the FK consumed that constraint, it did not create it. The FK's only footprint on the referenced side is PostgreSQL's internal RI trigger, dropped automatically with the referencing table. Four other tables (`deploy_history`, `deploy_queue`, `agent_keypairs`, `docker_container_events`) still reference `managed_hosts(name)`, so the constraint is load-bearing regardless. A fresh build's `managed_hosts` is identical with or without this pair.
- **Nothing in `src/` reads the table or the channel.** No non-test source references `stack_status` as a table or `stack_change` as a NOTIFY channel; `stack-status-broadcast-service.ts` listens on `docker_container_change` and `deploy_change`. The `stack_status_error` matches are an unrelated SSE event name, and the two `stack_change` matches are test titles in `startup-recovery.test.ts` / `deploy-watchdog.test.ts`, describing the broadcast concept rather than the pg channel.

### Deletion rather than renumbering

Renumbering `013_stack_status.sql` was rejected for two independent reasons:

1. `migrate.ts` tracks applied migrations by exact filename (`SELECT 1 FROM migrations WHERE name = $1`). A rename is a new name, so the renamed file re-runs on the live database, re-creating a table that was deliberately dropped.
2. `migrate.ts` sorts lexicographically, so any number above `019` would place the create after `019_drop_stack_status.sql`. A fresh database would end up with `stack_status` present, which is exactly the state the drop existed to prevent.

### Why deleting the drop is safe here

Images publish on every push to main (`type=sha` plus `latest`), so every commit between the two merges is a deployable version in which `stack_status` exists and has not yet been dropped. A database sitting in that window would keep the table forever once `019` is gone from the tree, since `migrate.ts` enumerates files on disk and would never see the drop again.

That window is empty: the only live database has already applied `019`. There is no deployment between `013_stack_status.sql` and `019_drop_stack_status.sql`, so the pair is inert everywhere it has run, and no database needs the drop preserved.

The alternative, had that not held, is to delete only `013_stack_status.sql` and keep `019` reduced to `DROP TABLE IF EXISTS stack_status;` plus `DROP FUNCTION IF EXISTS notify_stack_change();` (dropping the table takes its trigger with it). That resolves the duplicate identically and stays a no-op on a fresh install. It is not needed here.

### Orphan ledger rows are inert

`migrate.ts` enumerates migration **files on disk** (`readdirSync(migrationsDir)`) and checks each one against the `migrations` table by exact filename. It never walks the table looking for files. A row naming a file that no longer exists is therefore never consulted: no error, no re-run, no drift. Deleting these files needs no database change.

**Optional cleanup, not required for correctness.** The live database keeps two `migrations` rows pointing at files that no longer exist. They are inert. If you would rather the ledger match the tree exactly:

```sql
DELETE FROM migrations WHERE name IN ('013_stack_status.sql', '019_drop_stack_status.sql');
```

Skipping this has no functional consequence.

## Flow

Two paths, neither of them runtime application code.

Migration application, unchanged in shape and two files shorter:

```
boot -> initServer -> runMigrations -> readdirSync(migrations/) -> sort() -> per file: SELECT 1 FROM migrations WHERE name = $1 -> apply + INSERT
```

The new guard, gating the same directory at test time:

```
bun test -> migration-filenames.test.ts -> readdirSync(MIGRATIONS_DIR) -> three-digit prefix shape + prefix uniqueness -> fail naming the offending files
```

The guard sits ahead of the `migrations` CI job rather than inside it: it needs no PostgreSQL, so it runs in the ordinary `bun test` job and catches a duplicate before a database is ever provisioned. It reads `MIGRATIONS_DIR` exported from `migrate.ts`, so it checks the directory the runner actually reads rather than a second copy of the path.

What the guard does not cover: it sees one tree. Two open PRs each adding `033_*.sql` are both green until one merges, and the second is re-checked only if its CI re-runs against the new base. Preventing that on the merge itself needs "Require branches to be up to date before merging" on the `Build, Test & License Check` context, or a merge queue. Without it the collision is caught by the `push: branches: [main]` run, after main is red rather than before.

## What else this touches

- **`scripts/migration-verify/fixture.ts` and `assertions.ts`.** `DROPPED_TABLES` held one entry, `'stack_status'`, and `assertSchema` looped over it asserting each table is absent. With the pair deleted that assertion would still pass, but for the wrong reason: absent because never created, rather than correctly dropped. Rather than leave a loop that cannot fail, the table assertion is tightened from `containsAll` to `sameSet`, which catches any surviving table permanently and for every table, not just the ones someone remembers to list. The public schema holds exactly the 14 tables the migrations create plus the ledger; continuous aggregates are views and materialization hypertables and chunks live in `_timescaledb_internal`, so an exact set is enforceable.
- **`scripts/migration-verify/run.ts` now fails on unacknowledged removals.** `removed` was computed and only logged, while `containsAll('head migrations are all recorded', ...)` tolerates any ledger superset. Deleting a migration that does not cancel out would have passed `upgrade` and only failed `fresh` if some assertion happened to cover the missing object. Removals now have to be named in `ACKNOWLEDGED_REMOVALS`.
- **The `migrations` CI job tolerates these two removed files.** In the `upgrade` scenario the merge-base runner creates then drops `stack_status`, then head's `runMigrations` applies nothing new: `containsAll('head migrations are all recorded', ...)` passes because the ledger is a superset of head's files, and `sameSet('upgrade applied exactly the new migrations', ...)` passes because both sides are empty. In `fresh` and `idempotent`, both of which count files on disk, the counts simply drop by two.
- **`CLAUDE.md` gotcha 23.** It said CI does not check and that main still carries the duplicate; both stop being true with this PR. Rewritten to state what the guard covers, what it cannot see, and that renaming or deleting an applied migration is a ledger event.
- **Nothing else.** No table, SSE channel, settings key, env var, or event type outside the two deleted files.
- **Migration numbers.** This PR adds no migration. After the deletion, `ls migrations/ | sed -E 's/^([0-9]+)_.*/\1/' | sort | uniq -d` returns nothing: `013` is now unique and `019` is unused. A gap at `019` is harmless; `migrate.ts` cares about sort order, not contiguity.

## Verification

- `bun run typecheck` clean (`tsc --noEmit`, no output). `tsconfig.json` includes `**/*.ts`, so the `scripts/migration-verify/` edits are covered.
- `bun test --isolate src/lib/database/`: 305 pass, 0 fail across 16 files. Run per-directory rather than as the full suite: local `bun --version` (1.3.11) is behind `.bun-version`, which makes the whole-suite result meaningless regardless of the diff.
- `ls migrations/ | sed -E 's/^([0-9]+)_.*/\1/' | sort | uniq -d` returns nothing.
- `git diff | grep -E "^\+\s*(//|\*|/\*)"` returns one four-line JSDoc on `ACKNOWLEDGED_REMOVALS`; no em or en dashes in the diff.

### The guard proven by mutation

The test is not trusted on a passing run alone. Each failure mode was triggered on the cleaned tree, observed, and the temporary file removed (`git status` confirms none is committed).

Restoring `migrations/013_stack_status.sql`, the actual historical defect:

```
+   "013 -> 013_drop_agent_token_columns.sql, 013_stack_status.sql",

(fail) migration filenames > no two files share a numeric prefix
```

`touch migrations/9_mutation_malformed.sql`:

```
+   "9_mutation_malformed.sql",

(fail) migration filenames > every file is three digits then an underscore, so lexicographic sort is numeric order
```

Both name the offending file, so the failure is actionable without going to look. A third test asserts the directory read returned at least one file, so a silently empty `readdirSync` cannot make the other two pass vacuously.

The collision test no longer keys a malformed filename to itself. It did, which meant a pair like `13_a.sql` / `13_b.sql` could never collide and the reported failure named the wrong problem; malformed names are now skipped and left to the test that owns them.

### Not verified locally

The `migrations` job needs a real TimescaleDB and there is no local PostgreSQL or Docker. The `fresh`, `upgrade`, and `idempotent` scenarios run only in CI. The `containsAll` to `sameSet` change on the table assertion is the one edit that could newly fail there, if some table exists in `public` that `EXPECTED_TABLES` does not list; the 14 `CREATE TABLE` statements across `migrations/` plus the ledger match the list exactly, but confirm against the job result on this PR.

## Summary by CodeRabbit

* **New Features**
  * Added validation to ensure migration filenames use unique three-digit numeric prefixes.
  * Strengthened schema verification to require an exact match of expected tables.
  * Added safeguards so only explicitly approved migration removals are accepted during upgrades.
* **Bug Fixes**
  * Removed obsolete stack-status table, trigger, and related cleanup from the migration set.
* **Documentation**
  * Updated migration “Gotchas” guidance with clearer rules on numbering and safe migration handling.

### CODE REVIEW

File: scripts/migration-verify/run.ts [hunk 1]
```diff
@@ -44,6 +44,12 @@ type MigrationRunner = (db: DatabaseClient) => Promise<void>;
 const REPO_ROOT = join(import.meta.dir, '..', '..');
 const DB_PREFIX = process.env.MIGRATION_DB_PREFIX ?? 'migverify';
 
+/**
+ * Every ledger check below tolerates a superset, so this list is all that
+ * separates a deliberate migration deletion from an accidental one.
+ */
+const ACKNOWLEDGED_REMOVALS = ['013_stack_status.sql', '019_drop_stack_status.sql'];
```
[the remark this answers is not in the document]
Comment (jaredglaser) [reply scripts/migration-verify/run.ts:51]: Correct, fixed in 87fc5f3. `sameSet('base revision applied its own migrations', ...)` and `sameSet('upgrade applied exactly the new migrations', ...)` are both exact-set, so the claim was wrong as written.

Went with a slightly different wording than the suggestion, to keep the reason a removal goes unnoticed rather than just the fact that it does:

```
/**
 * A removed migration keeps its ledger row and no check below flags it, so this
 * list is all that separates a deliberate deletion from an accidental one.
 */
```

---
_Generated by [Claude Code](https://claude.ai/code)_
